tests/testsuite/artifact_dep.rs

```diff
@@ -0,0 +1,2049 @@
+        .with_stderr_contains("[COMPILING] bar v0.5.0 ([CWD]/bar)")
+        .with_stderr_contains("[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]")
+        .run();
+
+    let build_script_output = build_script_output_string(&p, "foo");
+    let msg = "we need the binary directory for this artifact along with all binary paths";
+    #[cfg(any(not(windows), target_env = "gnu"))]
+    {
+        match_exact(
+            "[..]/artifact/bar-[..]/bin/baz-[..]\n\
+             [..]/artifact/bar-[..]/staticlib/libbar-[..].a\n\
+             [..]/artifact/bar-[..]/cdylib/[..]bar.[..]\n\
+             [..]/artifact/bar-[..]/bin\n\
+             [..]/artifact/bar-[..]/bin/bar-[..]\n\
+             [..]/artifact/bar-[..]/bin/bar-[..]",
+            &build_script_output,
+            msg,
+            "",
+            None,
+        )
+        .unwrap();
+    }
+    #[cfg(all(windows, not(target_env = "gnu")))]
+    {
+        match_exact(
+            "[..]/artifact/bar-[..]/bin/baz.exe\n\
+             [..]/artifact/bar-[..]/staticlib/bar-[..].lib\n\
+             [..]/artifact/bar-[..]/cdylib/bar.dll\n\
+             [..]/artifact/bar-[..]/bin\n\
+             [..]/artifact/bar-[..]/bin/bar.exe\n\
+             [..]/artifact/bar-[..]/bin/bar.exe",
+            &build_script_output,
+            msg,
+            "",
+            None,
+        )
+        .unwrap();
+    }
+
+    assert!(
+        !p.bin("bar").is_file(),
+        "artifacts are located in their own directory, exclusively, and won't be lifted up"
+    );
+    assert!(!p.bin("baz").is_file(),);
+    assert_artifact_executable_output(&p, "debug", "bar", "bar");
+}
+
+#[cargo_test]
+fn build_script_with_bin_artifact_and_lib_false() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.0"
+                authors = []
+                resolver = "2"
+                
+                [build-dependencies]
+                bar = { path = "bar/", artifact = "bin" }
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+            fn main() {
+               bar::doit()
+            }
+        "#,
+        )
+        .file("bar/Cargo.toml", &basic_bin_manifest("bar"))
+        .file("bar/src/main.rs", "fn main() { bar::doit(); }")
+        .file(
+            "bar/src/lib.rs",
+            r#"
+            pub fn doit() {
+               panic!("cannot be called from build script due to lib = false");
+            }
+        "#,
+        )
+        .build();
+    p.cargo("build -Z unstable-options -Z bindeps")
+        .masquerade_as_nightly_cargo()
+        .with_status(101)
+        .with_stderr_contains(
+            "error[E0433]: failed to resolve: use of undeclared crate or module `bar`",
+        )
+        .with_stderr_contains(" --> build.rs:3:16")
```

Comment:
Just a heads up, we try to avoid matching against the output of `rustc` as much as possible.  If the output ever changes, then it requires doing awkward updates (because the cargo tests need to pass in the `rust` repository, but cargo is a submodule, so it can't be updated simultaneously).

This doesn't necessarily need to change, just wanted to caution against doing it too much.

Reply:
Oh, thanks for the heads-up. I understand and reviewed the artifact tests for this, adjusting two out of three in 78f0add6f6871601f886ed40e109027a7bfd0b27 to not depend on rustc anymore.

Only one remains which sets up a test where I couldn't use the absence of panics for assertions as I can't reliably run the artifact that's built for a different target. It's possible to run that on some platforms, but not all, by using `  cross_compile::can_run_on_host()`, but I considered that to not be strong enough. What would remain there is to not assert on the rustc output at all (_"error[E0425]: cannot find function `f2` in crate `d2`\n --> d1/src/main.rs:6:17"_), which leaves the test wide open for false negatives though.